text/0000-futures.md

```diff
@@ -0,0 +1,813 @@
+
+```rust
+/// A way of waking up a specific task.
+///
+/// Any task executor must provide a way of signaling that a task it owns
+/// is ready to be `poll`ed again. Executors do so by implementing this trait.
+pub trait Wake: Send + Sync {
+    /// Indicates that the associated task is ready to make progress and should
+    /// be `poll`ed.
+    ///
+    /// Executors generally maintain a queue of "ready" tasks; `wake` should place
+    /// the associated task onto this queue.
+    fn wake(&Arc<self>);
+}
+```
+
+Note that this trait uses `Arc` and hence is only available in `std`; however,
+it's a convenience on top of an `UnsafeWake` trait we'll see in the `no_std`
+section.
+
+In general futures are not coupled to any particular executor, so we use a trait
+object to handle waking:
+
+```rust
+/// A `Waker` is a handle for waking up a task by notifying its executor that it
+/// is ready to be run.
+///
+/// This handle contains a trait object pointing to an instance of the `UnsafeWake`
+/// trait, allowing notifications to get routed through it.
+pub struct Waker { ... }
+
+impl Waker {
+    /// Wake up the task associated with this `Waker`.
+    pub fn wake(&self);
+}
+
+impl Clone for Waker { .. }
+
+// We will see how to handle the no_std case later in the RFC...
+impl<T> From<Arc<T>> for Waker where T: Wake + 'static { ... }
+```
+
+Task execution always happens in the context of a `Waker` that can be used to
+wake the task up; we'll see the full `core::task::Context` structure below.
+
+### Executors
+
+An executor is responsible for polling tasks to completion. We represent this
+with the `core::task::BoxExecutor` trait (more on the name below):
+
+```rust
+/// A task executor.
+///
+/// A *task* is a `()`-producing future that runs at the top level, and will
+/// be `poll`ed until completion. It's also the unit at which wake-up
+/// notifications occur. Executors, such as thread pools, allow tasks to be
+/// spawned and are responsible for putting tasks onto ready queues when
+/// they are woken up, and polling them when they are ready.
+pub trait BoxExecutor {
+    /// Spawn the given task, polling it until completion.
+    ///
+    /// # Errors
+    ///
+    /// The executor may be unable to spawn tasks, either because it has
+    /// been shut down or is resource-constrained.
+    fn spawn(&mut self, task: Task) -> Result<(), SpawnError>;
+
+    /// Determine whether the executor is able to spawn new tasks.
+    ///
+    /// # Returns
+    ///
+    /// An `Ok` return means the executor is *likely* (but not guaranteed)
+    /// to accept a subsequent spawn attempt. Likewise, an `Err` return
+    /// means that `spawn` is likely, but not guaranteed, to yield an error.
+    fn status(&self) -> Result<(), SpawnError> {
+        Ok(())
+    }
+}
+
+pub struct Task { .. }
+
+// this impl is in `std` only:
+impl From<Box<dyn Future<Item = ()> + Send>> for Task {
+    pub fn from_box(task: ) -> Task;
```

Comment:
missing argument type? or maybe an indication that the argument type isn't interesting (aka `_`) ?